Jetpack Cloud: Provide Redux store

tyxla opened the pull request:
It was brought up by both @enejb and @rcoll that the Jetpack Cloud lacks access to the Redux store. AFAIU, the Redux store is of paramount importance to continue building the Jetpack Cloud, so I drafted this PR to address that.

#### Changes proposed in this Pull Request

* Add support to common boot for registering router after boot has setup all middleware
* Move Jetpack Cloud route registration later, so context middleware can be setup before that
* Wrap the Jetpack Cloud main layout in a `Provider` to pass store down and allow for `connect()` calls in subcomponents
* Display the current user's username as a demonstration of the Redux store.

A positive side effect here is that all the middleware is not only called now, but is also setup before the routes, so in Jetpack Cloud you now have access to the current user, locale, and (almost) everything you have in Calypso by default. Some of it may make sense to be disabled in future PRs, of course.

#### Testing instructions

* Checkout this branch.
* Run `CALYPSO_ENV=jetpack-cloud-development npm start`
* Go to http://jetpack.cloud.localhost:3000/scan
* Verify you can see your username in the content of the Scan page:

![](https://cldup.com/_X9DGKUdO4.png)

* Run `npm start`
* Verify Calypso still works well without any errors.

tyxla committed:
Boot: Add support for additional routing

tyxla committed:
Jetpack Cloud: Init router after boot

tyxla committed:
Jetpack Cloud: Wrap layout in Provider

tyxla committed:
Jetpack Cloud: Display the current user username using Redux

delawski commented:
Thank you for putting this together, @tyxla!

The code looks good, however, when testing locally, I noticed that the "Dashboard" link redirects to `/read` now. The other routes like `/backups` or `/scan` work as expected.

I guess this may have something to do with the fact that now the Cloud router is initialized after the default Calypso routes are defined so that the "Reader" takes precedence.

Other than that everything works well. I did a quick test and inserted a `<CurrentSite />` component at the top of the sidebar menu. It got rendered as expected:

<img width="403" alt="Screenshot 2020-02-18 at 10 43 53" src="https://user-images.githubusercontent.com/478735/74723976-d4811980-523b-11ea-93cd-7e25f8b7c8bb.png">

tyxla committed:
Boot: Disable loggedInMiddleware for Jetpack Cloud environments

tyxla commented:
Good catch @delawski! Fixed by https://github.com/Automattic/wp-calypso/pull/39501/commits/e4ea5da0d437f1405cf48d7956536cd8a0445ba0. 

The problem is that the root route is still handled by the reader section, and the logged-in middleware will create the redirect to `/read`.

This should be made easier to fix by #39460, but in the meantime, I've added an exception for the Jetpack Cloud environments and added a TODO comment to remember to remove it 😉 

This PR is ready for another 👀

delawski reviewed (approved):
Thank you for addressing the issue. Now everything works as expected 👍

rcoll reviewed (approved):
Thanks @tyxla ! Code looks fine and this works as expected in both test scenarios.
